### ISSUE TITLE
`NPE` on `HttpResponse` callback In gRPC client

### ISSUE THREAD

**Issue by Lincong** [line/armeria#4852]
## TLDR
I am on 1.16.2 and using gRPC client with multiple decorators. 
I recently added a new decorator `delegate.execute(ctx, req).mapError(...)` as a temporary workaround to handle [this issue](https://github.com/line/armeria/issues/4835). However, I encountered a `NPE` issue when the client retries. I suspect it is a bug.

## Specifics
My original plan was to reproduce the issue in a unit test on my fork of Armeria 1.16.2 and then create this issue with all information required to reproduce the issue (because that should be the best way to report a potential bug). However, I have not successfully reproduced it due to the fact that we have several decorators and a non-trivial amount of customization logic (>.<). Then I decided to create this issue and share the information I have at this point anyway.

## Decorator structure

```scala
// Decorator 1: exception reporting decorator
clientBuilder.decorator((delegate, ctx, req) => {
   delegate.execute(ctx, req).mapHeaders(respHeaders => {
      // If the response header contains certain information about exception, throw here.
      // This is our customized exception propagation from server to client.
      throw new StatusRuntimeException(...)
   }
}

// Decorator 2: circuit break decorator
clientBuilder.decorator(CircuitBreakerClient.newDecorator(...))

// Decorator 3: concurrency limit decorator
val concurrencyLimit: ConcurrencyLimit = ...
val concurrencyLimitingDecorator = ConcurrencyLimitingClient.newDecorator(concurrencyLimit)
clientBuilder.decorator(concurrencyLimitingDecorator)

// Decorator 4: this is added as a recommended workaround for: https://github.com/line/armeria/issues/4835
clientBuilder.decorator((delegate, ctx, req) => {
   delegate.execute(ctx, req).mapError(t => {
        // Copied from `com.linecorp.armeria.internal.client.ClientUtil.fail(ClientRequestContext ctx, Throwable cause)`
	val request: HttpRequest = ctx.request
	if (request != null) request.abort(t)
	val logBuilder: RequestLogBuilder = ctx.logBuilder
	logBuilder.endRequest(t)
	logBuilder.endResponse(t)
	t // continue to propagate the exception.
   }
}

// Decorator 5: add a callback to `RequestLogAccess`
clientBuilder.decorator((delegate, ctx, req) => {
   // I expect this callback to always be invoked when an exception is thrown from decorator 1 or 2.
   // Please let me know if this expectation is reasonable.
   ctx.log().whenComplete().thenAccept(reqLog => { /* A callback */ })
   delegate.execute(ctx, req)
}

// Decorator 6: retry
val mapping: RetryConfigMapping = ...
clientBuilder.decorator(RetryingClient.builderWithMapping(mapping).newDecorator())
```

## NPE stack trace
Posted in the comment section.

## Observation
1. When an exception is throw from `decorator 1`, retry happens (as expected).
2. After a few times of retrying, `NPE` happens.
3. Without `decorator 4`, everything is fine (no NPE).
4. Everything is fine (no NPE), if `decorator 4` is modified as below:

```scala
// Add a no-op `mapData` callback on `HttpResponse`.
clientBuilder.decorator((delegate, ctx, req) => {
   delegate.execute(ctx, req).mapError(t => {
        // Copied from `com.linecorp.armeria.internal.client.ClientUtil.fail(ClientRequestContext ctx, Throwable cause)`
	val request: HttpRequest = ctx.request
	if (request != null) request.abort(t)
	val logBuilder: RequestLogBuilder = ctx.logBuilder
	logBuilder.endRequest(t)
	logBuilder.endResponse(t)
	t // continue to propagate the exception.
   }.mapData(d => d) // Add this no-op callback to avoid NPE <----- HERE
```

**Comment by Lincong** [comment]
### NPE stack trace
- Please search for `java.lang.NullPointerException`.
- We shade our Armeria dependencies. So please ignore `grpc_shaded` package prefix.
- It feels like a bug of how `FuseableStreamMessage` is set up and/or invoked. 
- My best hope: Armeria OSS maintainers can read these stack trace, understand the issue/bug immediately, and provide some insights/explaination :)

```
grpc_shaded.io.grpc.StatusRuntimeException: UNKNOWN
at grpc_shaded.io.grpc.Status.asRuntimeException(Status.java:535)
at grpc_shaded.io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:533)
at grpc_shaded.com.linecorp.armeria.internal.client.grpc.ArmeriaClientCall.close(ArmeriaClientCall.java:506)
at grpc_shaded.com.linecorp.armeria.internal.client.grpc.ArmeriaClientCall.transportReportStatus(ArmeriaClientCall.java:445)
at grpc_shaded.com.linecorp.armeria.internal.common.grpc.HttpStreamDeframer.processOnError(HttpStreamDeframer.java:145)
at grpc_shaded.com.linecorp.armeria.internal.common.stream.DecodedStreamMessage$DecodingSubscriber.onError(DecodedStreamMessage.java:274)
at grpc_shaded.com.linecorp.armeria.common.stream.DeferredStreamMessage$ForwardingSubscriber.onError(DeferredStreamMessage.java:398)
at grpc_shaded.com.linecorp.armeria.internal.common.stream.AbortedStreamMessage.subscribe0(AbortedStreamMessage.java:91)
at grpc_shaded.com.linecorp.armeria.internal.common.stream.AbortedStreamMessage.subscribe(AbortedStreamMessage.java:83)
at grpc_shaded.com.linecorp.armeria.common.stream.DeferredStreamMessage.safeOnSubscribeToUpstream(DeferredStreamMessage.java:300)
at grpc_shaded.com.linecorp.armeria.common.stream.DeferredStreamMessage.delegate(DeferredStreamMessage.java:151)
at grpc_shaded.com.linecorp.armeria.common.DeferredHttpResponse.delegate(DeferredHttpResponse.java:47)
at grpc_shaded.com.linecorp.armeria.common.DeferredHttpResponse.lambda$delegateWhenComplete$0(DeferredHttpResponse.java:58)
at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:836)
at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:811)
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
at grpc_shaded.com.linecorp.armeria.client.retry.RetryingClient.handleRetryDecision(RetryingClient.java:432)
at grpc_shaded.com.linecorp.armeria.client.retry.RetryingClient.lambda$handleResponseWithoutContent$4(RetryingClient.java:325)
at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:836)
at java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:848)
at java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2168)
at java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:110)
at grpc_shaded.com.linecorp.armeria.client.retry.RetryingClient.handleResponseWithoutContent(RetryingClient.java:323)
at grpc_shaded.com.linecorp.armeria.client.retry.RetryingClient.lambda$doExecute0$3(RetryingClient.java:301)
at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:836)
at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:811)
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990)
at grpc_shaded.com.linecorp.armeria.common.util.UnmodifiableFuture.doCompleteExceptionally(UnmodifiableFuture.java:167)
at grpc_shaded.com.linecorp.armeria.common.util.UnmodifiableFuture.lambda$wrap$0(UnmodifiableFuture.java:101)
at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:836)
at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:811)
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
at grpc_shaded.com.linecorp.armeria.common.stream.DeferredStreamMessage.lambda$delegate$0(DeferredStreamMessage.java:132)
at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:836)
at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:811)
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
at grpc_shaded.com.linecorp.armeria.common.stream.StreamMessageCollector.onComplete(StreamMessageCollector.java:67)
at grpc_shaded.com.linecorp.armeria.common.stream.AbstractStreamMessage$CloseEvent.notifySubscriber(AbstractStreamMessage.java:258)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberOfCloseEvent0(DefaultStreamMessage.java:300)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberOfCloseEvent(DefaultStreamMessage.java:292)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.handleCloseEvent(DefaultStreamMessage.java:429)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber0(DefaultStreamMessage.java:372)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber(DefaultStreamMessage.java:328)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.addObjectOrEvent(DefaultStreamMessage.java:314)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.close(DefaultStreamMessage.java:436)
at grpc_shaded.com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.closeAction(HttpResponseDecoder.java:335)
at grpc_shaded.com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.cancelTimeoutOrLog(HttpResponseDecoder.java:370)
at grpc_shaded.com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.close(HttpResponseDecoder.java:318)
at grpc_shaded.com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.close(HttpResponseDecoder.java:308)
at grpc_shaded.com.linecorp.armeria.client.Http2ResponseDecoder.onHeadersRead(Http2ResponseDecoder.java:216)
at grpc_shaded.com.linecorp.armeria.client.Http2ResponseDecoder.onHeadersRead(Http2ResponseDecoder.java:225)
at grpc_shaded.io.netty.handler.codec.http2.Http2FrameListenerDecorator.onHeadersRead(Http2FrameListenerDecorator.java:48)
at grpc_shaded.io.netty.handler.codec.http2.Http2EmptyDataFrameListener.onHeadersRead(Http2EmptyDataFrameListener.java:63)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:409)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:337)
at grpc_shaded.io.netty.handler.codec.http2.Http2InboundFrameLogger$1.onHeadersRead(Http2InboundFrameLogger.java:56)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader$2.processFragment(DefaultHttp2FrameReader.java:476)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.readHeadersFrame(DefaultHttp2FrameReader.java:484)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:253)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:159)
at grpc_shaded.io.netty.handler.codec.http2.Http2InboundFrameLogger.readFrame(Http2InboundFrameLogger.java:41)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:173)
at grpc_shaded.io.netty.handler.codec.http2.DecoratingHttp2ConnectionDecoder.decodeFrame(DecoratingHttp2ConnectionDecoder.java:63)
at grpc_shaded.io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:393)
at grpc_shaded.io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:453)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
at grpc_shaded.io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:280)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
at grpc_shaded.io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1373)
at grpc_shaded.io.netty.handler.ssl.SslHandler.decodeNonJdkCompatible(SslHandler.java:1247)
at grpc_shaded.io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1287)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
at grpc_shaded.io.netty.handler.flush.FlushConsolidationHandler.channelRead(FlushConsolidationHandler.java:152)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
at grpc_shaded.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at grpc_shaded.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
at grpc_shaded.io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
at grpc_shaded.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
at grpc_shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
at grpc_shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
at grpc_shaded.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
at grpc_shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
at grpc_shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
at grpc_shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
at java.lang.Thread.run(Thread.java:748)
Cause: java.lang.NullPointerException:
at grpc_shaded.com.linecorp.armeria.common.stream.FuseableStreamMessage.lambda$collect$0(FuseableStreamMessage.java:162)
at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:836)
at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:811)
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
at grpc_shaded.com.linecorp.armeria.common.stream.DeferredStreamMessage.lambda$delegate$0(DeferredStreamMessage.java:132)
at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:836)
at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:811)
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
at grpc_shaded.com.linecorp.armeria.common.stream.StreamMessageCollector.onComplete(StreamMessageCollector.java:67)
at grpc_shaded.com.linecorp.armeria.common.stream.AbstractStreamMessage$CloseEvent.notifySubscriber(AbstractStreamMessage.java:258)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberOfCloseEvent0(DefaultStreamMessage.java:300)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberOfCloseEvent(DefaultStreamMessage.java:292)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.handleCloseEvent(DefaultStreamMessage.java:429)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber0(DefaultStreamMessage.java:372)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber(DefaultStreamMessage.java:328)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.addObjectOrEvent(DefaultStreamMessage.java:314)
at grpc_shaded.com.linecorp.armeria.common.stream.DefaultStreamMessage.close(DefaultStreamMessage.java:436)
at grpc_shaded.com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.closeAction(HttpResponseDecoder.java:335)
at grpc_shaded.com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.cancelTimeoutOrLog(HttpResponseDecoder.java:370)
at grpc_shaded.com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.close(HttpResponseDecoder.java:318)
at grpc_shaded.com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.close(HttpResponseDecoder.java:308)
at grpc_shaded.com.linecorp.armeria.client.Http2ResponseDecoder.onHeadersRead(Http2ResponseDecoder.java:216)
at grpc_shaded.com.linecorp.armeria.client.Http2ResponseDecoder.onHeadersRead(Http2ResponseDecoder.java:225)
at grpc_shaded.io.netty.handler.codec.http2.Http2FrameListenerDecorator.onHeadersRead(Http2FrameListenerDecorator.java:48)
at grpc_shaded.io.netty.handler.codec.http2.Http2EmptyDataFrameListener.onHeadersRead(Http2EmptyDataFrameListener.java:63)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:409)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:337)
at grpc_shaded.io.netty.handler.codec.http2.Http2InboundFrameLogger$1.onHeadersRead(Http2InboundFrameLogger.java:56)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader$2.processFragment(DefaultHttp2FrameReader.java:476)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.readHeadersFrame(DefaultHttp2FrameReader.java:484)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:253)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:159)
at grpc_shaded.io.netty.handler.codec.http2.Http2InboundFrameLogger.readFrame(Http2InboundFrameLogger.java:41)
at grpc_shaded.io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:173)
at grpc_shaded.io.netty.handler.codec.http2.DecoratingHttp2ConnectionDecoder.decodeFrame(DecoratingHttp2ConnectionDecoder.java:63)
at grpc_shaded.io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:393)
at grpc_shaded.io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:453)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
at grpc_shaded.io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:280)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
at grpc_shaded.io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1373)
at grpc_shaded.io.netty.handler.ssl.SslHandler.decodeNonJdkCompatible(SslHandler.java:1247)
at grpc_shaded.io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1287)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468)
at grpc_shaded.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
at grpc_shaded.io.netty.handler.flush.FlushConsolidationHandler.channelRead(FlushConsolidationHandler.java:152)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
at grpc_shaded.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
at grpc_shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at grpc_shaded.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
at grpc_shaded.io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
at grpc_shaded.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
at grpc_shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
at grpc_shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
at grpc_shaded.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
at grpc_shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
at grpc_shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
at grpc_shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
at java.lang.Thread.run(Thread.java:748)
```

**Comment by ikhoon** [comment]
I checked that `FuseableStreamMessage.collect()` throws a NPE when `mapError` is applied to a normal `StreamMesage`.
```java
StreamMessage.of(1, 2, 3, 4)
             .mapError(IllegalStateException::new)
             // Raises a NullPointerException
             .collect().join();
```
Let me send a patch for this.
